### PR TITLE
Move `FileCompressionType` out of `common` and into `core`

### LINKED ISSUE #7516: Move `FileCompressionType` out of `common` and into `core`
### Is your feature request related to a problem or challenge?

`FileCompressionType` appears to be the only item in `common` that uses several heavy dependencies in the `common` crate such as `tokio`. As result causes it to be a dependency of all crates using `common`, such as `expr`. Other than for `FileCompressionType` there is no usage of async in any crate except `core` _(tests excluded)_

By moving `FileCompressionType`, or the associated `convert_X` functions to `core`, we'd remove several heavy dependencies from all crates except core

### Describe the solution you'd like

A quick look over the codebase, it seems that `FileCompressionType` is only used in `core`. So it seems natural that it could easily be moved in to `core`. 

### Describe alternatives you've considered

- Make all async dependencies optional & control them via feature flags. 
- Keep FileCompressionType in `common`, but move the `convert_X` functions in to `core`. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by haohuaijin**
## Which issue does this PR close?

Closes #7516

## What changes are included in this PR?

- move `FileCompressionType` from `common` to `core/datasource/file_compression_type.rs`.
- because `FileType`'s `get_ext_with_compression` method is related to `FileCompressionType` and `get_ext_with_compression` method is only used in `core`, so I also move `get_ext_with_compression` to `core` by add a trait `FileTypeExt`.

### CODE REVIEW

File: datafusion-cli/Cargo.lock [hunk 1]
```diff
@@ -1124,20 +1124,11 @@ version = "31.0.0"
 dependencies = [
  "arrow",
  "arrow-array",
- "async-compression",
```
Comment (alamb) [inline datafusion-cli/Cargo.lock:1127]: 💯 -- FYI @universalmind303

File: datafusion/common/Cargo.toml [hunk 2]
```diff
@@ -35,34 +35,19 @@ path = "src/lib.rs"
...
-compression = ["xz2", "bzip2", "flate2", "zstd", "async-compression"]
-default = ["compression", "parquet"]
+default = ["parquet"]
 pyarrow = ["pyo3", "arrow/pyarrow"]
 
 [dependencies]
 apache-avro = { version = "0.15", default-features = false, features = ["snappy"], optional = true }
 arrow = { workspace = true }
 arrow-array = { workspace = true }
-async-compression = { version = "0.4.0", features = ["bzip2", "gzip", "xz", "zstd", "futures-io", "tokio"], optional = true }
```
Comment (alamb) [inline datafusion/common/Cargo.toml:46]: This is a very nice reduction in dependencies

File: datafusion/core/src/datasource/file_format/file_compression_type.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+                return Err(DataFusionError::NotImplemented(
+                    "Compression feature is not enabled".to_owned(),
+                ))
+            }
+            UNCOMPRESSED => Box::new(r),
+        })
+    }
+}
+
+/// Trait for extending the functionality of the `FileType` enum.
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/file_compression_type.rs:228]: this is perfect

FYI @devinjdangelo I think you originally consolidated the FileType enum -- this now splits out the dependency heavy part into `datasource`

**Review by alamb — APPROVED** [review]
Thank you very much @haohuaijin  👏  very nice

**Comment by alamb** [comment]
Thanks again @haohuaijin

**Comment by jonmmease** [comment]
Thanks for the fix @haohuaijin! This prevented the ability to compile `datafusion-common` to WASM, which VegaFusion depends on.

**Pull request merged**